<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">

    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print" />

    <title>Design-creation by htgt</title>
  </head>

  <body>

    <header>
      <div class="container">
        <h1>Design-creation</h1>
        <h2></h2>

        <section id="downloads">
          <a href="https://github.com/htgt/Design-Creation/zipball/master" class="btn">Download as .zip</a>
          <a href="https://github.com/htgt/Design-Creation/tarball/master" class="btn">Download as .tar.gz</a>
          <a href="https://github.com/htgt/Design-Creation" class="btn btn-github"><span class="icon"></span>View on GitHub</a>
        </section>
      </div>
    </header>

    <div class="container">
      <section id="main_content">
        <h1>DESIGN-CREATION</h1>

<p>Create a Knock-Out design for a given target.</p>

<p>Overview:</p>

<ol>
<li>Specify target area(s) coordinates for design.</li>
<li>Produce oligo target ( candidate ) region sequences.</li>
<li>Produce list of possible oligos, using AOS.</li>
<li>Filter and rank oligos, pick the best one of each type.</li>
<li>Find Valid U/D Region Oligo Pairs ( conditional designs only )</li>
<li>Find Best Gap Oligo Pair</li>
<li>Consolidate Design Data</li>
<li>Persist design with oligos to LIMS2.</li>
</ol><hr><h1>INPUT DATA</h1>

<h2>Design Parameters:</h2>

<ul>
<li>target coordinates

<ul>
<li>species</li>
<li>genome / assembly</li>
<li>chromosome name</li>
<li>chromosome strand</li>
<li>chromosome coordiantes
OR</li>
</ul>
</li>
<li>target gene

<ul>
<li>species</li>
<li>genomes / assembly</li>
<li>start exon</li>
<li>end exon
PLUS</li>
</ul>
</li>
<li>design type</li>
<li>phase - conditional</li>
<li>name</li>
</ul><h2>Oligo Profile:</h2>

<ul>
<li>oligo parameters</li>
<li>define region oligos must be located in</li>
</ul><h3>Conventional Deletion / Insertion</h3>

<pre><code>   G5               U5       TARGET       D3               G3
|======|---------|======|-|          |-|======|---------|======|
|------|                                                           &gt; G5 Region Length
       |------------------|                                        &gt; G5 Region Offset
                 |------|                                          &gt; U5 Region Length
                        |-|                                        &gt; U5 Region Offset
                                       |------|                    &gt; D3 Region Length
                                     |-|                           &gt; D3 Region Offset
                                                        |------|   &gt; G3 Region Length
                                     |------------------|          &gt; G3 Region Offset
</code></pre>

<h3>Conventional Conditional</h3>

<pre><code>   G5              U-REGION           D-REGION             G3
|======|---------|==========|-------|==========|--------|======|
|------|                                                           &gt; G5 Region Length
       |---------|                                                 &gt; G5 Region Offset
                      |-|                                          &gt; U Oligo Min Gap 
                    |-----|                                        &gt; U Oligo Max Gap 
                                         |-|                       &gt; D Oligo Min Gap 
                                       |-----|                     &gt; D Oligo Max Gap 
                                                        |------|   &gt; G3 Region Length
                                             |-----------------|   &gt; G3 Region Offset
</code></pre>

<hr><h1>WORKFLOW</h1>

<h1>1: Specify Target Area</h1>

<p>Given a target gene with details of the exons that are being targetted we need to produce
a set of coordinates for the target region(s);</p>

<h3>input:</h3>

<ul>
<li>target gene</li>
<li>start exon</li>
<li>end exon</li>
<li>species</li>
<li>assembly</li>
</ul><h3>output:</h3>

<ul>
<li>design coordinates</li>
</ul><hr><h1>2: Oligo Target Regions</h1>

<p>NOTE: Current program start at this step</p>

<p>Given the target coordinates for the design plus the parameters for the oligos
produce oligo target region sequence files ( input to aos ).
For deletion designs just one target region ( region to delete ).</p>

<p>For knockout designs we will have 2 ( U region and D region ).
Need to split these regions in half to get a region for each oligo.</p>

<h3>input:</h3>

<ul>
<li>design params

<ul>
<li>design coordinates

<ul>
<li>chromosome</li>
<li>strand</li>
<li>start</li>
<li>end</li>
</ul>
</li>
<li>species</li>
<li>assembly</li>
</ul>
</li>
<li>oligo profile</li>
</ul><h3>output: sequence for target region of each oligo</h3>

<ul>
<li>G5 region sequence</li>
<li>U5 region sequence</li>
<li>U3 region sequence ( optional )</li>
<li>D5 region sequence ( optional )</li>
<li>D3 region sequence</li>
<li>G3 region sequence</li>
</ul><p>Sequence files must have fasta headers formated as such:</p>

<pre><code>&gt;oligo_name:start-end
eg
&gt;U5:123124-123234
</code></pre>

<hr><h1>3: Produce Oligos</h1>

<p>Run AOS multiple times for each oligo, group and validate AOS output if needed.</p>

<h3>input:</h3>

<ul>
<li>design type</li>
<li>location of target region sequence files</li>
</ul><h3>output:</h3>

<ul>
<li>list of oligos for every target region</li>
</ul><h2>AOS Wrapper</h2>

<p>Takes target sequence and output list of primers for that region.
Need to wrap up input and output for aos to produce usable output for next step.</p>

<p>To work out assembly coordinates we use the offset information returned by aos for each oligo.
This information plus the oligo target region coordinates is enough to work out
the assembly coordinates for the oligos.</p>

<h3>input:</h3>

<ul>
<li>query sequence file</li>
<li>target sequence file ( currently chr file )</li>
<li>AOS parameters

<ul>
<li>oligo length ( 50 )</li>
<li>minimum GC content ( 28 )</li>
<li>number oligos ( 3 )</li>
<li>mask by lower-case? ( no )</li>
<li>genomic search method ( blat )</li>
</ul>
</li>
</ul><h3>output:</h3>

<ul>
<li>list of oligos, with following details:

<ul>
<li>sequence</li>
<li>assembly coordinates</li>
<li>id</li>
</ul>
</li>
</ul><hr><h1>4: Oligo Filtering</h1>

<p>We will have multiple oligos of each type, need to filter out bad ones.</p>

<h3>input:</h3>

<ul>
<li>oligos</li>
<li>design type</li>
</ul><h3>output:</h3>

<ul>
<li>validated oligo for each oligo type</li>
<li>throw error if we can't find given oligo type</li>
</ul><h2>Oligo Validity</h2>

<p>Check following:</p>

<ul>
<li>Coordinates for oligo are worked out correctly</li>
<li>Length of oligo is correct</li>
<li>Sequence is the same as the sequence from ensembl for given coordinates</li>
</ul><h2>Oligo Specificity</h2>

<p>inside bac ( but use genomes flanking sequence, 100k by default )</p>

<hr><h1>5: Find Valid U/D Region Oligo Pairs ( conditional designs )</h1>

<p>The pair of oligos from the U and D regions of conditional designs were produced
by splitting the respective regions in two and getting the appropriate oligo from each half.</p>

<p>We must find the best pairs of U and D oligos that fit the gap criteria for those oligos.
They may have a minimum gap distance between the oligos, and a optional maximum gap distance too.</p>

<p>The closer the oligos are to the minimum gap distance the better.</p>

<h3>input:</h3>

<ul>
<li>U5 and U3 oligos</li>
<li>D5 and D3 oligos</li>
<li>U Oligo Min Gap</li>
<li>U Oligo Max Gap ( optional )</li>
<li>D Oligo Min Gap ( optional )</li>
<li>D Oligo Max Gap</li>
</ul><h3>output</h3>

<ul>
<li>Best pair of U5 and U3 oligos</li>
<li>Best pair of D5 and D3 oligos</li>
</ul><hr><h1>6: Find Best Gap Oligo Pair</h1>

<p>We need to carry out additional checks on the G5 and G3 oligos and then
find the best pair of oligos to use.</p>

<p>All the oligo sequences and tiled into sections of sequence 6 bases long and then
compared against each other. If a pair of oligos has any matching 6 base sequences
then it is not a good oligo pair.</p>

<h3>input:</h3>

<ul>
<li>G5 and G3 oligos</li>
</ul><h3>output</h3>

<ul>
<li>Best pair of G5 and G3 oligos</li>
</ul><hr><h1>7: Consolidate Design Data</h1>

<p>Create the yaml design file, that will be used in the next persist step.
Brings together all the data from the previous steps and formats it correctly.
Also calls seperate module to work out phase of design.</p>

<h3>input:</h3>

<ul>
<li>target name</li>
<li>species</li>
<li>design type</li>
<li>oligos

<ul>
<li>type</li>
<li>sequence</li>
<li>loci

<ul>
<li>assembly</li>
<li>chr name</li>
<li>start</li>
<li>end</li>
</ul>
</li>
</ul>
</li>
<li>created by</li>
</ul><h3>output</h3>

<ul>
<li>design data yaml file</li>
</ul><h2>Design Phase</h2>

<p>To work our the phase the target transcript and oligo coordinates are required.</p>

<hr><h1>8: Persist Design</h1>

<p>Once we have valid oligos for the target we persist it, to LIMS2.
Use the LIMS2 api to insert design.</p>

<h3>input:</h3>

<ul>
<li>design data yaml file</li>
</ul><h3>output:</h3>

<ul>
<li>design stored in LIMS2</li>
</ul>
      </section>
    </div>

    
  </body>
</html>